runtime/src/sandbox.c

```diff
@@ -94,58 +55,70 @@ sandbox_allocate_stack(struct sandbox *sandbox)
 	assert(sandbox);
 	assert(sandbox->module);
 
-	int rc = 0;
+	return wasm_stack_allocate(&sandbox->stack, sandbox->module->stack_size);
+}
 
-	char *addr = mmap(NULL, sandbox->module->stack_size + /* guard page */ PAGE_SIZE, PROT_NONE,
-	                  MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
-	if (unlikely(addr == MAP_FAILED)) {
-		perror("sandbox allocate stack");
-		goto err_stack_allocation_failed;
-	}
+static inline void
+sandbox_free_stack(struct sandbox *sandbox)
+{
+	assert(sandbox);
 
-	/* Set the struct sandbox, HTTP Req/Resp buffer, and the initial Wasm Pages as read/write */
-	char *addr_rw = mmap(addr + /* guard page */ PAGE_SIZE, sandbox->module->stack_size, PROT_READ | PROT_WRITE,
-	                     MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED, -1, 0);
-	if (unlikely(addr_rw == MAP_FAILED)) {
-		perror("sandbox set stack read/write");
-		goto err_stack_allocation_failed;
+	return wasm_stack_free(&sandbox->stack);
+}
+
+/**
+ * Allocate http request and response buffers for a sandbox
+ * @param sandbox sandbox that we want to allocate HTTP buffers for
+ * @returns 0 on success, -1 on error
+ */
+static inline int
+sandbox_allocate_http_buffers(struct sandbox *self)
+{
+	self->request = vec_u8_new(self->module->max_request_size);
+	if (self->request == NULL) return -1;
+
+	self->response = vec_u8_new(self->module->max_response_size);
+	if (self->response == NULL) {
+		vec_u8_free(self->request);
+		return -1;
 	}
 
-	sandbox->stack.start = addr_rw;
-	sandbox->stack.size  = sandbox->module->stack_size;
+	return 0;
+}
 
-	rc = 0;
-done:
-	return rc;
-err_stack_prot_failed:
-	rc = munmap(addr, sandbox->stack.size + PAGE_SIZE);
-	if (rc == -1) perror("munmap");
-err_stack_allocation_failed:
-	sandbox->stack.start = NULL;
-	sandbox->stack.size  = 0;
-	goto done;
+static inline struct sandbox *
+sandbox_allocate(void)
+{
+	struct sandbox *sandbox                   = NULL;
+	size_t          page_aligned_sandbox_size = round_up_to_page(sizeof(struct sandbox));
+	sandbox                                   = calloc(1, page_aligned_sandbox_size);
+	sandbox_set_as_allocated(sandbox);
+	return sandbox;
 }
 
 /**
- * Allocates a new sandbox from a sandbox request
- * Frees the sandbox request on success
- * @param sandbox_request request being allocated
- * @returns sandbox * on success, NULL on error
+ * Allocates HTTP buffers and performs our approximation of "WebAssembly instantiation"
+ * @param sandbox
+ * @returns 0 on success, -1 on error
  */
-struct sandbox *
-sandbox_allocate(struct sandbox_request *sandbox_request)
+int
+sandbox_prepare_execution_environemnt(struct sandbox *sandbox)
```

Comment:
The semantics here are a bit of a mess. The "execution environment" is a webassembly module instance. It doesn't totally line up because this also allocates the http buffers.